### DISCUSSION TITLE
nuxt/kit: using createResolver to generate paths relatively to buildDir to improve type safety with docker

### DISCUSSION THREAD

**Discussion by ExileofAranei** [nuxt/nuxt#30383] [Ideas]
If a custom module needs to inject a type template there is a conventional function `createResolver` from `@nuxt/kit` package for resolving paths.

The result of `createResolver(import.meta.url).resolve(...)` is an absolute path from base dir. There are no problems with such paths on local machine, but using docker and its volumes the type safety is partially gone due to the absolute paths from docker container could not be resolved locally. The `.nuxt` and `node_modules` folders come from docker container.

For example, using [nuxt-auth](https://github.com/sidebase/nuxt-auth/blob/52d4b9a5b48431eb9d522865410d209c6d66a613/src/module.ts#L207) or [nuxt-auth-sanctum](https://github.com/manchenkoff/nuxt-auth-sanctum/blob/81a4ed05da7b7713825a2c075cc8e6e0a62081ae/src/templates.ts#L7-L8) modules, the auto generated files contain such imports:
```typescript
// Local machine:
// types/auth-misc.d.ts
import { RouteOptions } from 'E:/frontend/nuxt-auth/src/runtime/types.ts'

// types/sanctum.d.ts
import type { SanctumAppConfig } from 'E:/development/nuxt-auth-sanctum/src/runtime/types/config.ts';
import type { SanctumGlobalMiddlewarePageMeta } from 'E:/development/nuxt-auth-sanctum/src/runtime/types/meta.ts';


// Docker
// types/auth-misc.d.ts
import { RouteOptions } from '/var/www/html/node_modules/@sidebase/nuxt-auth/dist/runtime/types.ts'

//  types/sanctum.d.ts
import type { SanctumAppConfig } from '/var/www/html/node_modules/nuxt-auth-sanctum/dist/runtime/types/config.ts';
import type { SanctumGlobalMiddlewarePageMeta } from '/var/www/html/node_modules/nuxt-auth-sanctum/dist/runtime/types/meta.ts';
```
Typescript could not resolve such imports from docker container locally.

Nuxt itself generates declaration files with paths relative to `<buildDir>`, such as `components.d.ts`, `imports.d.ts`, which causes absolutely no problem using with docker, type imports are resolved correctly:
```typescript
import type { DefineComponent, SlotsType } from 'vue'
type IslandComponent<T extends DefineComponent> = T & DefineComponent<{}, {refresh: () => Promise<void>}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, SlotsType<{ fallback: { error: unknown } }>>
interface _GlobalComponents {
    'AppDialog': typeof import("../components/AppDialog.vue")['default']
    'AppLoader': typeof import("../components/AppLoader.vue")['default']
    'AppLogo': typeof import("../components/AppLogo.vue")['default']
    'AppOverlay': typeof import("../components/AppOverlay.vue")['default']
    // ...
```

Should nuxt provide a helper function for resolving such imports relatively to buildDir by default to improve the type safety even if docker is used? I've made a [PR](https://github.com/manchenkoff/nuxt-auth-sanctum/pull/273) for `nuxt-auth-sanctum` module to fix that, maybe there is a more elegant way to resolve such issues in a nuxt-way.

**Comment by manchenkoff** [comment]
Would really appreciate @danielroe's feedback on that, it might be a useful insight for module authors
